### PR TITLE
Added matrix exponentiation approach for finding fibonacci number.

### PR DESCRIPTION
**Pull request by mahbubcseju**
- Implemented matrix exponentiation for finding nth fibonnaci.
- Complexity is about O(nlog(n)*8), here 8 is for matrix multiplication of size 2 by 2.

### CODE REVIEW

File: matrix/nth_fibonnacci_using_matrix_exponentiation.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+def nth_fibonnaci_test(n):
+    if n <= 1:
+        return n
+    fib0 = 0
+    fib1 = 1
+    for i in range(2, n + 1):
+        tem = fib1 + fib0
+        fib0 = fib1
+        fib1 = tem
```
Comment (cclauss) [inline matrix/nth_fibonnacci_using_matrix_exponentiation.py]: In Python, these three lines could be replaced one line with:
```python
fib0, fib1 = fib1, fib0 + fib1
```
The right hand side will get fully evaluated before the variable assignment happens on the left hand side.

Comment (mahbubcseju) [reply matrix/nth_fibonnacci_using_matrix_exponentiation.py]: Resolved . Thanks  a lot .

File: matrix/nth_fibonnacci_using_matrix_exponentiation.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+As we know
+    f[n] = f[n-1] + f[n-1]
+Converting to matrix,
+    [f(n),f(n-1)] = [[1,1],[1,0]] * [f(n-1),f(n-2)]
+->  [f(n),f(n-1)] = [[1,1],[1,0]]^2 * [f(n-2),f(n-3)]
+    ...
+    ...
+->  [f(n),f(n-1)] = [[1,1],[1,0]]^(n-1) * [f(1),f(0)]
+
+So we just need the n times multiplication of  the matrix [1,1],[1,0]].
```
Comment (cclauss) [inline matrix/nth_fibonnacci_using_matrix_exponentiation.py]: Please remove the extra space.

File: matrix/nth_fibonnacci_using_matrix_exponentiation.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+We can decrease the n times multiplication by following  the divide and conquer approach
```
Comment (cclauss) [inline matrix/nth_fibonnacci_using_matrix_exponentiation.py]: Please remove the extra space.

What is the time complexity of the brute force method?  Which one is faster?  How much faster?

File: matrix/nth_fibonnacci_using_matrix_exponentiation.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        matrix_c.append(list_1)
+    return matrix_c
+
+
+def identity(n):
+    return [[int(row == column) for column in range(n)] for row in range(n)]
+
+
+def zerro(n):
+    return [[int(row == column) for column in range(n)] for row in range(n)]
```
Comment (cclauss) [inline matrix/nth_fibonnacci_using_matrix_exponentiation.py]: Can we lose this unused function?

File: matrix/nth_fibonnacci_using_matrix_exponentiation.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+
+def nth_fibonacci(n):
```
Comment (cclauss) [inline matrix/nth_fibonnacci_using_matrix_exponentiation.py]: Please add some doctests of the form:
```python
    def nth_fibonacci(n):
        """
        >>> nth_fibonacci(100)
        354224848179261915075
        >>> nth_fibonacci(-100)
        -100
        """
```
And the run __python -m doctests -v matrix/nth_fibonnacci_using_matrix_exponentiation.py__ to make sure that they pass.  Also consider adding doctest to other key functions.  Our automated testing will also run these tests on each pull request.

File: matrix/nth_fibonnacci_using_matrix_exponentiation.py (hunk 2 above)
Comment (mahbubcseju) [reply matrix/nth_fibonnacci_using_matrix_exponentiation.py]: removed

File: matrix/nth_fibonnacci_using_matrix_exponentiation.py (hunk 3 above)
Comment (mahbubcseju) [reply matrix/nth_fibonnacci_using_matrix_exponentiation.py]: removed and added

File: matrix/nth_fibonnacci_using_matrix_exponentiation.py (hunk 4 above)
Comment (mahbubcseju) [reply matrix/nth_fibonnacci_using_matrix_exponentiation.py]: removed

File: matrix/nth_fibonnacci_using_matrix_exponentiation.py (hunk 5 above)
Comment (mahbubcseju) [reply matrix/nth_fibonnacci_using_matrix_exponentiation.py]: Added and tested locally.

**Comment by cclauss** [comment]
```python
from timeit import timeit
print(timeit("nth_fibonacci(1000)", "from __main__ import nth_fibonacci", number=10000))
print(timeit("nth_fibonacci_test(1000)", "from __main__ import nth_fibonacci_test", number=10000))
```
1.565457206
1.273862879

**Comment by mahbubcseju** [comment]
> ```python
> from timeit import timeit
> print(timeit("nth_fibonacci(1000)", "from __main__ import nth_fibonacci", number=10000))
> print(timeit("nth_fibonacci_test(1000)", "from __main__ import nth_fibonacci_test", number=10000))
> ```
> 
> 1.565457206
> 1.273862879

Yes . If the input is high like n=100000, you will get the difference.

**Comment by mahbubcseju** [comment]
> ```python
> from timeit import timeit
> print(timeit("nth_fibonacci(1000)", "from __main__ import nth_fibonacci", number=10000))
> print(timeit("nth_fibonacci_test(1000)", "from __main__ import nth_fibonacci_test", number=10000))
> ```
> 
> 1.565457206
> 1.273862879

from timeit import timeit
print(timeit("nth_fibonacci(1000000)", "from __main__ import nth_fibonacci", number=5))
print(timeit("nth_fibonacci_test(1000000)", "from __main__ import nth_fibonacci_test", number=5))
2.3342058970001744
57.256506615000035

File: matrix/nth_fibonacci_using_matrix_exponentiation.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+As we know
+    f[n] = f[n-1] + f[n-1]
+Converting to matrix,
+    [f(n),f(n-1)] = [[1,1],[1,0]] * [f(n-1),f(n-2)]
+->  [f(n),f(n-1)] = [[1,1],[1,0]]^2 * [f(n-2),f(n-3)]
+    ...
+    ...
+->  [f(n),f(n-1)] = [[1,1],[1,0]]^(n-1) * [f(1),f(0)]
+So we just need the n times multiplication of  the matrix [1,1],[1,0]].
+We can decrease the n times multiplication by following  the divide and conquer approach.
```
Comment (cclauss) [inline matrix/nth_fibonacci_using_matrix_exponentiation.py]: `<space><space> --> <space>` on these two lines.

File: matrix/nth_fibonacci_using_matrix_exponentiation.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    n = n - 1
+    while n > 0:
+        if n % 2 == 1:
+            res_matrix = multiply(res_matrix, fibonacci_matrix)
+        fibonacci_matrix = multiply(fibonacci_matrix, fibonacci_matrix)
+        n = int(n / 2)
+    return res_matrix[0][0]
+
+
+def nth_fibonacci_test(n):
```
Comment (cclauss) [inline matrix/nth_fibonacci_using_matrix_exponentiation.py]: Looking good... Please add a doctest here.

**Comment by cclauss** [comment]
Perhaps the functions should be renamed to clarify intent:
* __nth_fibonacci()__ --> __nth_fibonacci_matrix()__
* __nth_fibonacci__test()__ --> __nth_fibonacci_bruteforce()__

File: matrix/nth_fibonacci_using_matrix_exponentiation.py (hunk 7 above)
Comment (mahbubcseju) [reply matrix/nth_fibonacci_using_matrix_exponentiation.py]: Added . And sorry for too much commit . Actually it is my first time contribution to any open source.

**Comment by mahbubcseju** [comment]
> * nth_fibonacci_bruteforce

Changed

**Pull request merged**